Departure Time Changes, replaces #34

Abby-Wheelis opened the pull request:
## Describe your changes
These patches enable departure time simulation in this library

## Issue ticket number and link
[#220](https://github.com/EVerest/EVerest/issues/220)

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

Abby-Wheelis committed:
port changes over from patches

These patches enable departure time simulation in this library

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
port changes over from patches

These patches enable departure time simulation in this library

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove unused variable

We will adjust the speed of simulation with future changes to the demo containers

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove unused variable

We will adjust the speed of simulation with future changes to the demo containers

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix whitespace issues

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix whitespace issues

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove whitespace

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove whitespace

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas commented on iso15118/evcc/states/iso15118_2_states.py:
```diff
@@ -63,6 +65,7 @@
     PowerDeliveryReq,
     PowerDeliveryRes,
     PreChargeReq,
+    ResponseCode,
```
Is not used

SebaLukas commented on iso15118/evcc/everest/ev_state.py:
```diff
@@ -12,7 +13,8 @@
     # Common
-    PaymentOption: str = ''
+    EAmount: int = 60
+    DepartureTime: Optional[int] = None
```
Why is DepartureTime optional? With the changes in everest-core `DepartureTime` is always set.

Abby-Wheelis replied:
I can make it non optional

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -295,7 +298,7 @@ async def get_charge_params_v2(self, protocol: Protocol) -> ChargeParamsV2:
             )
             ac_charge_params = ACEVChargeParameter(
-                departure_time=0,
+                departure_time=EVEREST_EV_STATE.DepartureTime,
```
If `DepartureTime` remains optional, then you must check here whether `DepartureTime` is `None`.

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -306,7 +309,7 @@ async def get_charge_params_v2(self, protocol: Protocol) -> ChargeParamsV2:
             )
             dc_charge_params = DCEVChargeParameter(
-                departure_time=0,
+                departure_time=EVEREST_EV_STATE.DepartureTime,
```
See above

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -422,7 +425,7 @@ async def get_scheduled_se_params(
 
         scheduled_params = ScheduledScheduleExchangeReqParams(
-            departure_time=7200,
+            departure_time=EVEREST_EV_STATE.DepartureTime,
```
See above

SebaLukas replied:
For D20 change back to the default value 7200

Abby-Wheelis replied:
Should the default value of 7200 override anything the user provided in D20? Or just use 7200 as the default?

Abby-Wheelis replied:
@SebaLukas Can you provide any clarity on if D20 requires overriding the user input to 7200, or is I should just make sure the default (absent input from the user) is 7200?

SebaLukas replied:
For D20 the default value should be 7200 for the time being. If the DepatureTime is also to be set for D20, this can be changed quickly.

SebaLukas replied:
This is not resolved. Please revert your change and write directly `7200`.

Abby-Wheelis replied:
This would be overriding all user input, not just providing a default value, is that what is required?

SebaLukas replied:
This is only for the ISO15118-20 messages. Not for the D2.
Right now there is no changes in the ISO15118-20 state machine for the `Departure Time` in `Josev` and  `libiso15118`.
Therefore, we should leave it as it was. As soon as there are changes for D20 that also affect departure time, the user input can be used.

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -436,7 +439,7 @@ async def get_dynamic_se_params(
         """Overrides EVControllerInterface.get_dynamic_se_params()."""
         dynamic_params = DynamicScheduleExchangeReqParams(
-            departure_time=7200,
+            departure_time=EVEREST_EV_STATE.DepartureTime,
```
See above

SebaLukas replied:
For D20 change back to the default value 7200

SebaLukas replied:
Same here

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -753,7 +774,7 @@ async def get_ac_charge_loop_params_v20(
             # Dynamic Mode
             dynamic_params = DynamicACChargeLoopReqParams(
-                departure_time=2000,
+                departure_time=EVEREST_EV_STATE.DepartureTime,
```
See above

SebaLukas replied:
Same here

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -905,7 +926,7 @@ async def get_charge_params_v2h(self, protocol: Protocol) -> ChargeParamsV2:
         )
         dc_charge_params = DCEVChargeParameter(
-            departure_time=0,
+            departure_time=EVEREST_EV_STATE.DepartureTime,
```
See above

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,161 @@
+@author: ANAND
+"""
+# KS is btwn 1 and 20
+def LQRChargeCurve(DepTime, EAmount, PMax, KS):
```
Define Return values: `-> `

SebaLukas replied:
Also applies to the remaining functions

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,161 @@
+import numpy as np
```
Is `numpy` really necessary? If `numpy` is essential, then add numpy to `requirements.txt` as well

Abby-Wheelis replied:
I can factor out numpy, it was just making matrix operations easier

Abby-Wheelis replied:
Actually, I think it would be best to put numpy back to keep the code readable. If the complexity of using `numpy` and `control` is a big concern, we can take out all complex calculations and just fulfill the charge request linearly (30 kW is 3 hours = 10 kW per hour ... more like 5 kW per 30 minutes due to limited number of data points we can put in the schedule).

SebaLukas replied:
If you need numpy then add your code back to the PR and add numpy to the `requirements.txt`. My only concern is that users don't just add Python packages like that.

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,161 @@
+import numpy as np
+import control as ct
```
Exactly the same as `numpy`. Is `control` really necessary or can it also be solved using existing modules?

Abby-Wheelis replied:
I don't think I can take this out, the function it uses are central to the calculation of the charge code, I'll add it to the requirements

SebaLukas replied:
This is okay if it is really necessary :+1:

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,161 @@
+            )
+        )
+    logger.debug(f'Dummy schedule: {temp}')
+    return temp
```
Missin EOF

SebaLukas reviewed (changes requested):
See my comments.
Please also check the Codacy  warnings.

Abby-Wheelis committed:
rmeove unused import

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
rmeove unused import

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
make departure time non optional and use 7200 as default

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
make departure time non optional and use 7200 as default

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
add EOF

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
add EOF

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
adding control to requirements

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
adding control to requirements

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove numpy

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove numpy

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas commented:
Check also the `Codacy` check. Some issues are "real" issues (Optional not used, etc.)

Abby-Wheelis commented:
What I am planning to do, for now, is keep `numpy` in, implement the simple curve solution, and take `control` out. This is because the dependencies of `control` interfere with the yocto build, and we need to think about another way to accomplish this without breaking the build. `numpy` doesn't cause any issues with the yocto build. I'll work an another branch and bring this to the working group meeting on Monday in case there are any other considerations I'm missing and the group favors a different approach.

Abby-Wheelis committed:
remove unused import

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove unused import

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix whitespace in power_curve

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix whitespace in power_curve

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
add numpy back to code base

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
add numpy back to code base

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas committed:
Adding ISO15118-20 AC support for the SIL (#35)

Signed-off-by: Sebastian Lukas <sebastian.lukas@pionix.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas committed:
Refactoring vars from iso15118_ev interface (#42)

Signed-off-by: Sebastian Lukas <sebastian.lukas@pionix.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

markxoe committed:
fix typing issue (#43)

Signed-off-by: Mark Oude Elberink <mark.oude.elberink@frickly.systems>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

andistorm committed:
Set deps for python targets (#44)

Adds iso15118_requirements_pip_install_dist as dep for install targets
of python package. This way the requirements.txt is installed
automatically and can't be forgotten.

Signed-off-by: Andreas Heinrich <andreas.heinrich@rwth-aachen.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
switch to simple linear algorithm

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
switch to simple linear algorithm

Abby-Wheelis committed:
remove library requirements

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove library requirements

Abby-Wheelis committed:
remove unused function

Abby-Wheelis committed:
remove unused function

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
update docstrings

Abby-Wheelis committed:
update docstrings

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
merge upstream

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix whitespace issues

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix whitespace issues

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
trailing whitespace

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
trailing whitespace

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas commented on iso15118/evcc/states/iso15118_2_states.py:
```diff
@@ -193,6 +196,7 @@ async def process_message(
             self.stop_state_machine("ChargeService not offered")
             return
 
+        logger.debug("hello -- about to select an auth option?")
```
Is this really needed after merging?

SebaLukas commented on iso15118/evcc/states/iso15118_2_states.py:
```diff
@@ -798,11 +805,19 @@ async def process_message(
+
+            # If end of profile > end of SECC schedule or no DT (dt==0), end renegotiation...
+            departure_time = EVEREST_EV_STATE.DepartureTime
+            if (departure_time is None or self.comm_session.end_of_profile_schedule >= departure_time or 0 == departure_time):
```
You can delete `departure_time is None`. `departure_time` is not optional anymore

SebaLukas reviewed (changes requested):
Look over my comments

SebaLukas commented:
DCO is not succesfull. Some commits are not sigend-off. If you want, you can rebase and squash all your commits in one commit.

Abby-Wheelis committed:
remove log statement

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove log statement

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
departure time will not be none

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
departure time will not be none

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis commented:
Once we come to a resolution on if we really need to override all user input to the departure time, I will rebase and squash my commits to fix DCO

shankari commented:
@SebaLukas can you please comment on the hardcoding versus the override so that we can finish this PR? If we write 7200, it will override the user input, which means that the departure time won't actually work, right?

Abby-Wheelis committed:
revert to 7200 until D20 states are implmented

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
revert to 7200 until D20 states are implmented

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Merge branch 'departure_time_signed' of https://github.com/Abby-Wheelis/ext-switchev-iso15118 into departure_time_signed

SebaLukas commented:
DCO still has a problem because a few commits are still signed off. In addition, the commits are duplicated.

Abby-Wheelis commented:
Yes, I'm still working out the best way to fix that issue, I think squashing as you mentioned might be what I need to do. I tried rebasing once as the DCO instructions in the check recommended, but that did not fix everything.

Abby-Wheelis commented:
I opened #45 to replace this after multiple failed rebase attempt that were affecting the commit history

SebaLukas commented:
To squash all 48 commits in one commit you can use:

```
git rebase -i HEAD~48
```
And mark all commits expect the first one with `s` in the text editor

Abby-Wheelis commented:
My rebase attempt led to dozens of merge conflicts and added changes to my branch that I did not make. I think the new PR is the best route to keep the commit history clean.
